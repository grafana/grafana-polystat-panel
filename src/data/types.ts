export const CUSTOM_SPLIT_DELIMITER = '#️⃣🔠🆗🆗🔠#️⃣';
