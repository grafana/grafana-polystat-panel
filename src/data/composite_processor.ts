import { includes as lodashIncludes } from 'lodash';
import { PolystatModel } from '../components/types';
import { getWorstSeries } from './threshold_processor';
import { ClickThroughTransformer } from './clickThroughTransformer';
import { stringToJsRegex, escapeStringForRegex, ScopedVars, InterpolateFunction, textUtil } from '@grafana/data';
import { getTemplateSrv } from '@grafana/runtime';
import { CompositeItemType, CompositeMetric } from '../components/composites/types';
import { CUSTOM_SPLIT_DELIMITER } from './types';

const resolveCompositeTemplates = (
  metricComposites: CompositeItemType[],
  replaceVariables: InterpolateFunction
): CompositeItemType[] => {
  const ret: CompositeItemType[] = [];
  metricComposites.forEach((item: CompositeItemType) => {
    const firstResolve = replaceVariables(item.name); // th e ScopedVars should expand here
    const resolved = replaceVariables(firstResolve, {}, customFormatter).split(CUSTOM_SPLIT_DELIMITER);
    resolved.forEach((newName: string) => {
      ret.push({
        ...item,
        name: newName,
        isTemplated: true,
        templatedName: item.name,
      });
    });
  });

  return ret;
};

const customFormatter = (value: any) => {
  if (Object.prototype.toString.call(value) === '[object Array]') {
    return value.join(CUSTOM_SPLIT_DELIMITER);
  }
  return value;
};

const resolveMemberTemplates = (
  compositeName: string,
  isTemplated: boolean,
  templatedName: string,
  members: CompositeMetric[],
  replaceVariables: InterpolateFunction
): any[] => {
  const ret: any[] = [];
  const variableRegex = /\$(\w+)|\[\[([\s\S]+?)(?::(\w+))?\]\]|\${(\w+)(?:\.([^:^\}]+))?(?::(\w+))?}/g;
  members.forEach((member) => {
    // Resolve templates in series names
    if (member.seriesMatch) {
      const matchResult = member.seriesMatch.match(variableRegex);
      if (matchResult && matchResult.length > 0) {
        matchResult.forEach((template) => {
          // if the template contains the composite template, replace it with the compositeName
          if (isTemplated && template.includes(templatedName)) {
            // replace it
            template = template.replace(templatedName, compositeName);
          }
          const resolvedSeriesNames = [replaceVariables(template, {}, 'raw')];
          resolvedSeriesNames.forEach((seriesName) => {
            const newName = member.seriesMatch.replace(template, seriesName);
            const escapedName = escapeStringForRegex(seriesName);
            const newSeriesNameEscaped = member.seriesMatch.replace(template, escapedName);
            ret.push({
              ...member,
              seriesName: newName,
              seriesNameEscaped: newSeriesNameEscaped,
            });
          });
        });
      } else {
        ret.push(member);
      }
    }
  });

  return ret;
};

const resolveMemberAliasTemplates = (name: string, matches: any): string => {
  const templateVars: ScopedVars = {};
  matches.forEach((name: string, i: number) => {
    templateVars[i] = { text: i, value: name };
  });
  if (matches.groups) {
    Object.keys(matches.groups).forEach((key) => {
      templateVars[key.replace(/\s+/g, '_')] = { text: key, value: matches.groups[key] };
    });
  }
  return getTemplateSrv().replace(name, templateVars);
};

/**
 * Copies values, leaves members empty
 */
const shallowClone = (item: PolystatModel): PolystatModel => {
  const clone: PolystatModel = {
    value: item.value,
    valueFormatted: item.valueFormatted,
    valueRounded: item.valueRounded,
    stats: item.stats,
    name: item.name,
    displayName: item.displayName,
    timestamp: item.timestamp,
    prefix: item.prefix,
    suffix: item.suffix,
    color: item.color,
    clickThrough: item.clickThrough,
    operatorName: item.operatorName,
    newTabEnabled: item.newTabEnabled,
    sanitizedURL: item.sanitizedURL,
    sanitizeURLEnabled: item.sanitizeURLEnabled,
    showName: item.showName,
    showValue: item.showValue,
    isComposite: item.isComposite,
    members: [],
  };
  return clone;
};

export const ApplyComposites = (
  composites: CompositeItemType[],
  data: PolystatModel[],
  replaceVariables: InterpolateFunction
): PolystatModel[] => {
  if (!composites) {
    return data;
  }
  const filteredMetrics: number[] = [];
  const clonedComposites: PolystatModel[] = [];
  // the composite Name can be a template variable
  // the composite should only match specific metrics or expanded templated metrics that use the composite name
  const resolvedComposites = resolveCompositeTemplates(composites, replaceVariables);
  for (let i = 0; i < resolvedComposites.length; i++) {
    const matchedMetrics: number[] = [];
    const aComposite = resolvedComposites[i];
    if (!aComposite.showComposite) {
      continue;
    }
    let currentWorstSeries = null;
    // this should filter the members that are matches for the composite name
    const templatedMembers = resolveMemberTemplates(
      aComposite.name,
      aComposite.isTemplated,
      aComposite.templatedName,
      aComposite.metrics,
      replaceVariables
    );
    for (let j = 0; j < templatedMembers.length; j++) {
      const aMetric = templatedMembers[j];
      // look for the matches to the pattern in the data
      for (let index = 0; index < data.length; index++) {
        // match regex
        // seriesName may not be defined yet, skip
        if (typeof aMetric.seriesMatch === 'undefined') {
          continue;
        }
        // name may not be escaped, check both
        let metricName = aMetric.seriesMatch;
        if (aMetric.seriesNameEscaped !== undefined) {
          metricName = aMetric.seriesNameEscaped;
        }
        const regex = stringToJsRegex(metricName);
        const matches = regex.exec(data[index].name);
        if (matches && matches.length > 0) {
          const seriesItem = data[index];
          // Template out the name of the metric using the alias
          if (aMetric.alias && aMetric.alias.length > 0) {
            seriesItem.displayName = resolveMemberAliasTemplates(aMetric.alias, matches);
          }

          // keep index of the matched metric
          matchedMetrics.push(index);
          // only hide if requested
          if (!aComposite.showMembers) {
            filteredMetrics.push(index);
          }
          if (aComposite.clickThrough && aComposite.clickThrough.length > 0) {
            // process template variables
            let url = replaceVariables(aComposite.clickThrough);
            // apply both types of transforms, one targeted at the data item index, and secondly the nth variant
            url = ClickThroughTransformer.transformComposite(aComposite.name, url);
            url = ClickThroughTransformer.transformSingleMetric(index, url, data);
            url = ClickThroughTransformer.transformNthMetric(url, data);
            seriesItem.clickThrough = url;
            seriesItem.sanitizedURL = textUtil.sanitizeUrl(url);
          }
        }
      }
    }
    if (matchedMetrics.length === 0) {
      continue;
    }
    // now determine the most triggered threshold

    for (let k = 0; k < matchedMetrics.length; k++) {
      const itemIndex = matchedMetrics[k];
      const seriesItem = data[itemIndex];
      // check thresholds
      if (currentWorstSeries === null) {
        currentWorstSeries = seriesItem;
      } else {
        currentWorstSeries = getWorstSeries(currentWorstSeries, seriesItem);
      }
    }
    // Prefix the valueFormatted with the actual metric name
    if (currentWorstSeries !== null) {
      const clone = shallowClone(currentWorstSeries);
      clone.name = aComposite.name;
      clone.displayName = aComposite.name;
      // tooltip/legend uses this to expand what values are inside the composite
      for (let index = 0; index < matchedMetrics.length; index++) {
        const itemIndex = matchedMetrics[index];
        clone.members.push({
          ...data[itemIndex],
          name: data[itemIndex].displayName || data[itemIndex].name,
        });
      }
      clone.thresholdLevel = currentWorstSeries.thresholdLevel;
      // currentWorstSeries.valueFormatted = currentWorstSeriesName + ': ' + currentWorstSeries.valueFormatted;
      // now push the composite into data
      // add the composite setting for showing the name/value to the new cloned model
      clone.showName = aComposite.showName;
      clone.showValue = aComposite.showValue;
      clone.displayMode = aComposite.displayMode;
      clone.newTabEnabled = aComposite.clickThroughOpenNewTab;
      clone.sanitizeURLEnabled = aComposite.clickThroughSanitize;
      // mark this series as a composite
      clone.isComposite = true;
      clonedComposites.push(clone);
    }
  }
  // now merge the clonedComposites into data
  Array.prototype.push.apply(data, clonedComposites);
  // sort by value descending
  filteredMetrics.sort((a, b) => {
    return b - a;
  });
  // now remove the filtered metrics from final list
  // remove filtered metrics, use splice in reverse order
  for (let i = data.length; i >= 0; i--) {
    if (lodashIncludes(filteredMetrics, i)) {
      data.splice(i, 1);
    }
  }
  return data;
};
