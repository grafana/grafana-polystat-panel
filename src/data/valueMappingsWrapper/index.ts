import { getMappedValue } from './v7/valueMappings';
import { getValueMappingResult } from './v8/valueMappings';

//
// To maintain compatibility back to Grafana >=v7.3, this wrapper is needed
// to process the value mappings based on the object passed in.
//
// the method `getMappedValue()` function was removed in v9
// the replacement `getValueMappingResult()` only appears in v8.1.0+
//
// At runtime, this cannot be determined so the legacy function and the new function are both provided
// here, to allow this to work work across all versions (7.3 > 9.x)
//
export const GetMappedValue = (valueMappings: any[], value: any): any => {
  let mappedValue = getMappedValue(valueMappings, value);
  if (typeof mappedValue === 'undefined') {
    return getValueMappingResult(valueMappings, value);
  }
  return mappedValue;
  // TODO: replace with this
  return getMappedValue(valueMappings, value) ?? getValueMappingResult(valueMappings, value);
};
