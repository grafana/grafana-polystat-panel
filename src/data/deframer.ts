import { DataFrame, Field, FieldType, ArrayVector, Labels } from '@grafana/data';
import _ from 'lodash';

// Inserts a "Time" field into each dataframe if it is missing
// the value of the timestamp is "now"
// any field without a numeric type is considered a label
export function InsertTime(data: DataFrame[]): DataFrame[] {
  // TODO: time to insert can be taken from the first row if there are timeseries already
  // for now, just insert now
  const timeToInsert = Date.now();
  const newData: DataFrame[] = [];
  for (let i = 0; i < data.length; i++) {
    const frame = data[i];
    //const flattened = this.flattenLabels(frame, 0);
    const newFrame = _.cloneDeep(frame);
    // clear the fields
    newFrame.fields = [];
    //const labels = this.getLabelsOfFrame(frame);
    const hasTimestamp = frameHasTimestamp(frame);
    // rebuild a new frame with labels on the numerical fields
    const numFields = frame.fields.length;
    for (let j = 0; j < numFields; j++) {
      const aField = frame.fields[j];
      if (aField.type === FieldType.number) {
        // need to get the number of rows of data for this frame
        const rowsOfField = aField.values.toArray().length;
        if (!hasTimestamp) {
          for (let rowNum = 0; rowNum < rowsOfField; rowNum++) {
            // only create a new field when the rowValue is not null
            if (aField.values.toArray()[rowNum] !== null) {
              // this has a nonnull value
              const flattened = flattenLabels(frame, rowNum);
              const newField = newFieldWithLabels(aField, flattened);
              const newFieldValues = new ArrayVector();
              const value = getValueOfField(aField, rowNum);
              newFieldValues.add(value);
              newField.values = newFieldValues;
              newFrame.fields.push(newField);
            }
          }
        } else {
          // copy the object
          const copiedField = Object.assign({}, aField);
          newFrame.fields.push(copiedField);
        }
      }
    }
    if (!hasTimestamp) {
      const z = new ArrayVector();
      z.add(timeToInsert);

      const timeField: Field = {
        name: 'Time',
        type: FieldType.time,
        values: z,
        config: null,
      };
      // insert it
      newFrame.fields.push(timeField);
    } else {
      // time field already exists
      // copy all non-number fields from original frame
      for (let j = 0; j < frame.fields.length; j++) {
        const aField = frame.fields[j];
        if (aField.type !== FieldType.number) {
          newFrame.fields.push(aField);
        }
      }
    }
    newData.push(newFrame);
  }

  return newData;
}

function frameHasTimestamp(frame: DataFrame): boolean {
  for (let j = 0; j < frame.fields.length; j++) {
    const aField = frame.fields[j];
    if (aField.type === FieldType.time) {
      return true;
    }
  }
  return false;
}

function flattenLabels(frame: DataFrame, rowNum: number) {
  let labelIndexes = [];
  const numFields = frame.fields.length;

  // first get the fields of type string
  for (let j = 0; j < numFields; j++) {
    if (frame.fields[j].type === FieldType.string) {
      labelIndexes.push(j);
    }
  }
  let labelWithValues = getLabelValues(frame, labelIndexes, rowNum);
  return labelWithValues;
}

function newFieldWithLabels(field: Field, labels: Labels): Field {
  const newField = _.cloneDeep(field);
  newField.labels = labels;
  return newField;
}

function getValueOfField(field: Field, index: number) {
  const bufferValue = field.values.toArray()[index];
  return bufferValue;
}

function getLabelValues(frame: DataFrame, indexes: any[], rowNum: number) {
  let labelAndValue = {};
  for (let index = 0; index < indexes.length; index++) {
    let indexValue = indexes[index];
    let aField = frame.fields[indexValue];
    if (aField.type !== FieldType.number) {
      let value = getValueOfField(aField, rowNum);
      labelAndValue[aField.name] = value;
    }
  }
  return labelAndValue;
}
