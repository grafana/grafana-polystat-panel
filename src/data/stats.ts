import { FieldCalcs } from '@grafana/data';
import { PolystatModel } from '../components/types';

export function GetValueByOperator(
  metricName: string,
  data: PolystatModel | null,
  operatorName: string,
  calcs: FieldCalcs
) {
  switch (operatorName) {
    case 'name':
      return metricName;
    case 'last_time':
      if (data) {
        return data.timestamp;
      } else {
        return Date.now();
      }
    default:
      let aValue = calcs[operatorName];
      return aValue;
  }
}
