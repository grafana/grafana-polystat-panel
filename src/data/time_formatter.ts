import { dateTime } from '@grafana/data';

export const TimeFormatter = (timeZone: string, timestamp: number, timestampFormat: string): string => {
  const timestampFormatted =
    timeZone === 'utc'
      ? dateTime(timestamp)
        .utc()
        .format(timestampFormat)
      : dateTime(timestamp).format(timestampFormat);
  return timestampFormatted;
}
