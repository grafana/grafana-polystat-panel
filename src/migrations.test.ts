import { PanelModel, RangeMap, ValueMap, SpecialValueMap } from '@grafana/data';
//import { CompositeItemType } from 'components/composites/types';
//import { OverrideItemType } from 'components/overrides/types';

import {
  PolystatPanelMigrationHandler,
  migrateDefaults,
  migrateComposites,
  migrateOverrides,
  AngularSavedComposites,
  AngularSavedOverrides,
  migrateValueAndRangeMaps,
} from './migrations';

describe('Polystat -> PolystatV2 migrations', () => {
  it('only migrates old polystat', () => {
    const panel = {} as PanelModel;
    const options = PolystatPanelMigrationHandler(panel);
    expect(options).toEqual({});
  });

  it('migrates old polystat config', () => {
    const panel = {} as PanelModel;
    panel.options = {};
    //@ts-ignore
    panel.polystat = {
      animationSpeed: 2500,
      columnAutoSize: true,
      columns: '6',
      defaultClickThrough: 'https://grafana.com',
      defaultClickThroughNewTab: false,
      defaultClickThroughSanitize: true,
      displayLimit: 100,
      ellipseCharacters: 18,
      ellipseEnabled: false,
      fontAutoColor: true,
      fontAutoScale: true,
      fontSize: 12,
      fontType: 'Roboto',
      globalDecimals: 2,
      globalDisplayMode: 'all',
      globalDisplayTextTriggeredEmpty: 'OK',
      globalOperatorName: 'avg',
      globalUnitFormat: 'short',
      gradientEnabled: true,
      hexagonSortByDirection: 1,
      hexagonSortByField: 'name',
      maxMetrics: 0,
      polygonBorderColor: '#000000',
      polygonBorderSize: 2,
      polygonGlobalFillColor: '#0a55a1',
      radius: '',
      radiusAutoSize: true,
      regexPattern: '',
      rowAutoSize: true,
      rows: '',
      shape: 'square',
      tooltipDisplayMode: 'all',
      tooltipDisplayTextTriggeredEmpty: 'OK',
      tooltipEnabled: true,
      tooltipFontSize: 12,
      tooltipFontType: 'Roboto',
      tooltipPrimarySortDirection: 2,
      tooltipPrimarySortField: 'thresholdLevel',
      tooltipSecondarySortDirection: 2,
      tooltipSecondarySortField: 'value',
      tooltipTimestampEnabled: true,
      valueEnabled: true,
    };
    const options = PolystatPanelMigrationHandler(panel);
    expect(options).toMatchSnapshot();
  });

  it('migrates old polystat config with mappings', () => {
    const panel = {
      options: {},
      mappingType: 1,
      rangeMaps: [
        {
          from: 'null',
          text: 'N/A',
          to: 'null',
        },
      ],
      valueMaps: [
        {
          op: '=',
          text: 'N/A',
          value: 'null',
        },
        {
          op: '=',
          text: 'Nominal',
          value: '30.386',
        },
      ],
    } as unknown as PanelModel;
    //@ts-ignore
    panel.polystat = {
      animationSpeed: 2500,
      columnAutoSize: true,
      columns: '6',
      defaultClickThrough: 'https://grafana.com',
      defaultClickThroughNewTab: false,
      defaultClickThroughSanitize: true,
      displayLimit: 100,
      ellipseCharacters: 18,
      ellipseEnabled: false,
      fontAutoColor: true,
      fontAutoScale: true,
      fontSize: 12,
      fontType: 'Roboto',
      globalDecimals: 2,
      globalDisplayMode: 'all',
      globalDisplayTextTriggeredEmpty: 'OK',
      globalOperatorName: 'avg',
      globalUnitFormat: 'short',
      gradientEnabled: true,
      hexagonSortByDirection: 1,
      hexagonSortByField: 'name',
      maxMetrics: 0,
      polygonBorderColor: '#000000',
      polygonBorderSize: 2,
      polygonGlobalFillColor: '#0a55a1',
      radius: '',
      radiusAutoSize: true,
      regexPattern: '',
      rowAutoSize: true,
      rows: '',
      shape: 'square',
      tooltipDisplayMode: 'all',
      tooltipDisplayTextTriggeredEmpty: 'OK',
      tooltipEnabled: true,
      tooltipFontSize: 12,
      tooltipFontType: 'Roboto',
      tooltipPrimarySortDirection: 2,
      tooltipPrimarySortField: 'thresholdLevel',
      tooltipSecondarySortDirection: 2,
      tooltipSecondarySortField: 'value',
      tooltipTimestampEnabled: true,
      valueEnabled: true,
    };
    const options = PolystatPanelMigrationHandler(panel);
    expect(options).toMatchSnapshot();
    expect(panel).toMatchSnapshot();
    expect(panel.fieldConfig.defaults.mappings[0].options.result.text).toEqual('N/A');
    expect(panel.fieldConfig.defaults.mappings[1].options).toEqual({ '30.386': { color: undefined, text: 'Nominal' } });
  });

  it('correctly converts top level config to new names', () => {
    const oldPolystatOptions = {
      polystat: {
        animationSpeed: 2500,
        columnAutoSize: true,
        columns: '6',
        defaultClickThrough: 'https://grafana.com',
        defaultClickThroughNewTab: false,
        defaultClickThroughSanitize: true,
        displayLimit: 100,
        ellipseCharacters: 18,
        ellipseEnabled: false,
        fontAutoColor: true,
        fontAutoScale: true,
        fontSize: 12,
        fontType: 'Roboto',
        globalDecimals: 2,
        globalDisplayMode: 'all',
        globalDisplayTextTriggeredEmpty: 'OK',
        globalOperatorName: 'avg',
        globalUnitFormat: 'short',
        gradientEnabled: true,
        hexagonSortByDirection: 1,
        hexagonSortByField: 'name',
        maxMetrics: 0,
        polygonBorderColor: '#000000',
        polygonBorderSize: 2,
        polygonGlobalFillColor: '#0a55a1',
        radius: '',
        radiusAutoSize: true,
        regexPattern: '',
        rowAutoSize: true,
        rows: '',
        shape: 'square',
        tooltipDisplayMode: 'all',
        tooltipDisplayTextTriggeredEmpty: 'OK',
        tooltipEnabled: true,
        tooltipFontSize: 12,
        tooltipFontType: 'Roboto',
        tooltipPrimarySortDirection: 2,
        tooltipPrimarySortField: 'thresholdLevel',
        tooltipSecondarySortDirection: 2,
        tooltipSecondarySortField: 'value',
        tooltipTimestampEnabled: true,
        valueEnabled: true,
      },
    };
    const options = migrateDefaults(oldPolystatOptions.polystat);
    console.log('top level...');
    console.log(JSON.stringify(options, null, 2));
    expect(options.layoutNumColumns).toEqual(6);
    expect(options.autoSizeColumns).toEqual(true);
    expect(options.globalClickthrough).toEqual('https://grafana.com');
    expect(options.compositeConfig.animationSpeed).toEqual('2500');
  });

  it('correctly migrates overrides', () => {
    const oldPolystatOptions: AngularSavedOverrides = {
      savedOverrides: [
        {
          clickThrough: '',
          colors: ['#299c46', '#e5ac0e', '#bf1b00', '#4040a0'],
          decimals: 3,
          enabled: true,
          label: 'OVERRIDE 1',
          metricName: '/.*TempInC.*/',
          newTabEnabled: true,
          operatorName: 'current',
          prefix: '',
          sanitizeURLEnabled: true,
          suffix: '',
          thresholds: [
            {
              color: '#299c46',
              state: 0,
              value: 0,
            },
            {
              color: '#e5ac0e',
              state: 1,
              value: 25,
            },
            {
              color: '#bf1b00',
              state: 2,
              value: 32,
            },
          ],
          unitFormat: 'celsius',
        },
        {
          clickThrough: '',
          colors: ['#299c46', '#e5ac0e', '#bf1b00', '#4040a0'],
          decimals: 3,
          enabled: true,
          label: 'OVERRIDE 2',
          metricName: '/.*TempInF.*/',
          newTabEnabled: true,
          operatorName: 'current',
          prefix: '',
          sanitizeURLEnabled: true,
          suffix: '',
          thresholds: [
            {
              color: '#299c46',
              state: 0,
              value: 0,
            },
            {
              color: '#e5ac0e',
              state: 1,
              value: 78,
            },
            {
              color: '#bf1b00',
              state: 2,
              value: 82,
            },
          ],
          unitFormat: 'fahrenheit',
        },
      ],
    };
    const options = migrateOverrides(oldPolystatOptions);
    console.log('overrides...');
    console.log(JSON.stringify(options, null, 2));
    expect(options.overrideConfig.overrides[0].label).toEqual('OVERRIDE 1');
  });

  it('correctly migrates composites to compositeConfig', () => {
    /*
    const panel = {
      options: {
        compositeConfig: {
          composites: [] as CompositeItemType[],
          enabled: true,
          animationSpeed: '500',
        },
      },
    } as PanelModel;
    */

    const oldPolystatOptions: AngularSavedComposites = {
      savedComposites: [
        {
          animateMode: 'all',
          clickThrough: '',
          compositeName: 'P1',
          displayName: '',
          enabled: true,
          hideMembers: true,
          label: 'COMPOSITE 1',
          members: [
            {
              seriesName: '/P1/',
            },
          ],
          newTabEnabled: true,
          sanitizeURLEnabled: true,
          sanitizedURL: '',
          showName: true,
          showValue: true,
          thresholdLevel: 0,
        },
        {
          animateMode: 'all',
          clickThrough: '',
          compositeName: 'P2',
          displayName: '',
          enabled: true,
          hideMembers: true,
          label: 'COMPOSITE 2',
          members: [
            {
              seriesName: '/P2/',
            },
          ],
          newTabEnabled: true,
          sanitizeURLEnabled: true,
          sanitizedURL: '',
          showName: true,
          showValue: true,
          thresholdLevel: 0,
        },
      ],
    };
    const options = migrateComposites(oldPolystatOptions, '2222');
    console.log('composites...');
    console.log(JSON.stringify(options, null, 2));
    expect(options.compositeConfig.animationSpeed).toEqual('2222');
  });

  it('correctly migrates range and value maps', () => {
    const aPanel = {
      mappingType: 1,
      valueMaps: [
        {
          op: '=',
          text: 'N/A',
          value: 'null',
        },
        {
          op: '=',
          text: 'Nominal',
          value: '30.386',
        },
      ],
      rangeMaps: [
        {
          from: 'null',
          text: 'N/A',
          to: 'null',
        },
        {
          from: '30',
          text: 'Nominal',
          to: '40',
        },
      ],
    };
    const newMaps = migrateValueAndRangeMaps(aPanel);
    console.log(JSON.stringify(newMaps, null, 2));
    expect(newMaps.length).toEqual(3);
    const aSpecialMap = newMaps[0] as SpecialValueMap;
    expect(aSpecialMap.type).toEqual('special');
    expect(aSpecialMap.options.result.text).toEqual('N/A');
    const aValueMap = newMaps[1] as ValueMap;
    expect(aValueMap.type).toEqual('value');
    expect(aValueMap.options).toEqual({ '30.386': { text: 'Nominal' } });
    const aRangeMap = newMaps[2] as RangeMap;
    expect(aRangeMap.type).toEqual('range');
    expect(aRangeMap.options.from).toEqual(30);
    expect(aRangeMap.options.to).toEqual(40);
    expect(aRangeMap.options.result.text).toEqual('Nominal');
  });
});

/*
        "defaultClickThrough": "",
        "defaultClickThroughNewTab": false,
        "defaultClickThroughSanitize": false,
        "displayLimit": 100,
        "ellipseCharacters": 18,
        "ellipseEnabled": false,
        "fontAutoColor": true,
        "fontAutoScale": true,
        "fontSize": 12,
        "fontType": "Roboto",
        "globalDecimals": 2,
        "globalDisplayMode": "all",
        "globalDisplayTextTriggeredEmpty": "OK",
        "globalOperatorName": "avg",
        "globalUnitFormat": "short",
        "gradientEnabled": true,
        "hexagonSortByDirection": 1,
        "hexagonSortByField": "name",
        "maxMetrics": 0,
        "polygonBorderColor": "#000000",
        "polygonBorderSize": 2,
        "polygonGlobalFillColor": "#0a55a1",
        "radius": "",
        "radiusAutoSize": true,
        "regexPattern": "",
        "rowAutoSize": true,
        "rows": "",
        "shape": "square",
        "tooltipDisplayMode": "all",
        "tooltipDisplayTextTriggeredEmpty": "OK",
        "tooltipEnabled": true,
        "tooltipFontSize": 12,
        "tooltipFontType": "Roboto",
        "tooltipPrimarySortDirection": 2,
        "tooltipPrimarySortField": "thresholdLevel",
        "tooltipSecondarySortDirection": 2,
        "tooltipSecondarySortField": "value",
        "tooltipTimestampEnabled": true,
        "valueEnabled": true
      },
*/
