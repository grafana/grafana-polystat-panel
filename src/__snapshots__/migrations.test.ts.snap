// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Polystat -> PolystatV2 migrations migrates old polystat config 1`] = `
Object {
  "autoSizeColumns": true,
  "autoSizePolygons": true,
  "autoSizeRows": true,
  "compositeConfig": Object {
    "animationSpeed": "2500",
    "composites": Array [],
    "enabled": true,
  },
  "ellipseCharacters": 18,
  "ellipseEnabled": false,
  "globalAutoScaleFonts": true,
  "globalClickthrough": "https://grafana.com",
  "globalClickthroughNewTabEnabled": false,
  "globalClickthroughSanitizedEnabled": true,
  "globalDecimals": 2,
  "globalDisplayMode": "all",
  "globalDisplayTextTriggeredEmpty": "OK",
  "globalFillColor": "#0a55a1",
  "globalFontSize": 12,
  "globalGradientsEnabled": true,
  "globalOperator": "mean",
  "globalPolygonBorderColor": "#000000",
  "globalPolygonBorderSize": 2,
  "globalPolygonSize": 50,
  "globalRegexPattern": "",
  "globalShape": "square",
  "globalShowValueEnabled": true,
  "globalTextFontAutoColor": "#000000",
  "globalTextFontAutoColorEnabled": true,
  "globalTextFontColor": "",
  "globalThresholdsConfig": Array [],
  "globalTooltipsEnabled": true,
  "globalTooltipsShowTimestampEnabled": true,
  "globalUnitFormat": "short",
  "layoutDisplayLimit": 100,
  "layoutNumColumns": 6,
  "layoutNumRows": 8,
  "overrideConfig": Object {
    "overrides": Array [],
  },
  "panelId": 0,
  "radius": 100,
  "sortByDirection": 1,
  "sortByField": "name",
  "tooltipDisplayMode": "all",
  "tooltipDisplayTextTriggeredEmpty": "OK",
  "tooltipPrimarySortByField": "thresholdLevel",
  "tooltipPrimarySortDirection": 2,
  "tooltipSecondarySortByField": "value",
  "tooltipSecondarySortDirection": 2,
}
`;

exports[`Polystat -> PolystatV2 migrations migrates old polystat config with mappings 1`] = `
Object {
  "autoSizeColumns": true,
  "autoSizePolygons": true,
  "autoSizeRows": true,
  "compositeConfig": Object {
    "animationSpeed": "2500",
    "composites": Array [],
    "enabled": true,
  },
  "ellipseCharacters": 18,
  "ellipseEnabled": false,
  "globalAutoScaleFonts": true,
  "globalClickthrough": "https://grafana.com",
  "globalClickthroughNewTabEnabled": false,
  "globalClickthroughSanitizedEnabled": true,
  "globalDecimals": 2,
  "globalDisplayMode": "all",
  "globalDisplayTextTriggeredEmpty": "OK",
  "globalFillColor": "#0a55a1",
  "globalFontSize": 12,
  "globalGradientsEnabled": true,
  "globalOperator": "mean",
  "globalPolygonBorderColor": "#000000",
  "globalPolygonBorderSize": 2,
  "globalPolygonSize": 50,
  "globalRegexPattern": "",
  "globalShape": "square",
  "globalShowValueEnabled": true,
  "globalTextFontAutoColor": "#000000",
  "globalTextFontAutoColorEnabled": true,
  "globalTextFontColor": "",
  "globalThresholdsConfig": Array [],
  "globalTooltipsEnabled": true,
  "globalTooltipsShowTimestampEnabled": true,
  "globalUnitFormat": "short",
  "layoutDisplayLimit": 100,
  "layoutNumColumns": 6,
  "layoutNumRows": 8,
  "overrideConfig": Object {
    "overrides": Array [],
  },
  "panelId": 0,
  "radius": 100,
  "sortByDirection": 1,
  "sortByField": "name",
  "tooltipDisplayMode": "all",
  "tooltipDisplayTextTriggeredEmpty": "OK",
  "tooltipPrimarySortByField": "thresholdLevel",
  "tooltipPrimarySortDirection": 2,
  "tooltipSecondarySortByField": "value",
  "tooltipSecondarySortDirection": 2,
}
`;

exports[`Polystat -> PolystatV2 migrations migrates old polystat config with mappings 2`] = `
Object {
  "fieldConfig": Object {
    "defaults": Object {
      "mappings": Array [
        Object {
          "options": Object {
            "match": "null",
            "result": Object {
              "color": undefined,
              "text": "N/A",
            },
          },
          "type": "special",
        },
        Object {
          "options": Object {
            "30.386": Object {
              "color": undefined,
              "text": "Nominal",
            },
          },
          "type": "value",
        },
      ],
    },
    "overrides": Array [],
  },
  "options": Object {},
}
`;
